packages/js/components/src/tree-select-control/index.js

```diff
@@ -0,0 +1,497 @@
+			leaves: {
+				/**
+				 * Return all leaf options flattened under this option. The options are filtered if in searching.
+				 *
+				 * @return {InnerOption[]} All leaf options that are flattened under this option. The options are filtered if in searching.
+				 */
+				get() {
+					if ( ! this.hasChildren ) {
+						return [];
+					}
+					return this.children.flatMap( ( option ) => {
+						return option.hasChildren ? option.leaves : option;
+					} );
+				},
+			},
+			checked: {
+				/**
+				 * Returns whether this option is checked.
+				 * A leaf option is checked if its value is selected.
+				 * A parent option is checked if all leaves are checked.
+				 *
+				 * @return {boolean} True if checked, false otherwise.
+				 */
+				get() {
+					if ( this.hasChildren ) {
+						return this.leaves.every( ( opt ) => opt.checked );
+					}
+					return cache.selectedValues.includes( this.value );
+				},
+			},
+			partialChecked: {
+				/**
+				 * Returns whether this option is partially checked.
+				 * A leaf option always returns false.
+				 * A parent option is partially checked if at least one but not all leaves are checked.
+				 *
+				 * @return {boolean} True if partially checked, false otherwise.
+				 */
+				get() {
+					if ( ! this.hasChildren ) {
+						return false;
+					}
+					return (
+						! this.checked &&
+						this.leaves.some(
+							( opt ) => opt.checked || opt.partialChecked
+						)
+					);
+				},
+			},
+			expanded: {
+				/**
+				 * Returns whether this option is expanded.
+				 * A leaf option always returns false.
+				 *
+				 * @return {boolean} True if expanded, false otherwise.
+				 */
+				get() {
+					return (
+						isSearching ||
+						this.value === ROOT_VALUE ||
+						cache.expandedValues.includes( this.value )
+					);
+				},
+			},
+		};
+
+		const reduceOptions = ( acc, { children = [], ...option } ) => {
+			if ( children.length ) {
+				option.children = children.reduce( reduceOptions, [] );
+
+				if ( ! option.children.length ) {
+					return acc;
+				}
+			} else if ( isSearching ) {
+				const match = option.label.toLowerCase().indexOf( filter );
+				if ( match === -1 ) {
+					return acc;
+				}
+				option.label = highlightOptionLabel( option.label, match );
+			}
+
+			Object.defineProperties( option, descriptors );
+			acc.push( option );
+
+			return acc;
+		};
+
+		const filteredTreeOptions = treeOptions.reduce( reduceOptions, [] );
+		cache.filteredOptionsMap.set( filter, filteredTreeOptions );
+
+		return filteredTreeOptions;
+	}, [ treeOptions, filter ] );
+
+	const onKeyDown = ( event ) => {
+		if ( disabled ) return;
+
+		if ( ESCAPE === event.key ) {
+			setTreeVisible( false );
+		}
+
+		if ( ENTER === event.key ) {
+			setTreeVisible( true );
+			event.preventDefault();
+		}
+
+		const stepDict = {
+			[ ARROW_UP ]: -1,
+			[ ARROW_DOWN ]: 1,
+		};
+		const step = stepDict[ event.key ];
+
+		if ( step && dropdownRef.current && filteredOptions.length ) {
```

Comment:
Can we compare `event.key` to `ARROW_DOWN` instead so we don't need `stepDict`?

Reply:
> In case there is not ARROW_UP nor ARROW_DOWN. Then step will be undefined and we don't execute the code.

Ah, thanks for the clarification. I'm good with this now.